libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1764,17 +1784,58 @@ bool PbftManager::pushPbftBlock_(PeriodData &&period_data, std::vector<std::shar
   // Advance pbft consensus period
   advancePeriod();
 
+  // Create new pillar block
+  // !!! Important: processPillarBlock must be called only after advancePeriod()
+  if (kGenesisConfig.state.hardforks.ficus_hf.isPillarBlockPeriod(pbft_period)) {
+    assert(pbft_period == pbft_chain_->getPbftChainSize());
+    processPillarBlock(pbft_period);
+  }
+
   return true;
 }
 
+void PbftManager::processPillarBlock(PbftPeriod current_pbft_chain_size) {
+  // Pillar block use state from current_pbft_chain_size - final_chain_->delegation_delay(), e.g. block with period 32
+  // uses state from period 27.
+  PbftPeriod request_period = current_pbft_chain_size - final_chain_->delegation_delay();
+  // advancePeriod() -> resetConsensus() -> waitForPeriodFinalization() makes sure block request_period was already
+  // finalized
+  assert(final_chain_->last_block_number() >= request_period);
+
+  const auto block_header = final_chain_->block_header(request_period);
+  const auto bridge_root = final_chain_->get_bridge_root(request_period);
+  const auto bridge_epoch = final_chain_->get_bridge_epoch(request_period);
```

Comment:
@kstdl it this change ok ? Doesn't t break something in contract ? Check `PbftPeriod request_period = current_pbft_chain_size - final_chain_->delegation_delay();`

Reply:
yeah. It should be just after that system transaction which is a case here if I am not mistaken